src/librustc_codegen_llvm/consts.rs

```diff
@@ -437,7 +451,7 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
                 //
                 // We could remove this hack whenever we decide to drop macOS 10.10 support.
                 if self.tcx.sess.target.target.options.is_like_osx {
-                    let sect_name = if alloc.bytes.iter().all(|b| *b == 0) {
+                    let sect_name = if alloc.alloc_content.is_none() {
```

Comment:
Is that change correct or should it check both conditions?

Reply:
this check checks for all zeros. so the old check was in fact "wrong" with respect to undef bytes (in quotes because we can choose to make undef bytes all zero). It was definitely wrong if the `alloc` contains relocations.

cc @RalfJung 	@eddyb 	we didn't make it hard enough to mis-use `Allocation`.

So we need a few things answered before we can solve this. Should undef bytes be treated as zero bytes here? Is bss cheaper than data? Only if we need all zeros anyway?

I'm assuming putting it in bss only makes sense if all bytes are initialized and zero, so imo the condition should be

```rust
alloc.get_bytes(&self.tcx, Pointer::new(AllocId(0), Size::ZERO), alloc.size).map_or(false, |b| b.iter().all(|b| *b == 0))
```

which makes sure there are no relocations and all bits are defined

Reply:
> cc @RalfJung @eddyb we didn't make it hard enough to mis-use Allocation.

I feel like I had raised exactly that concern when `Allocation` had most of its internals become public. :/

`alloc.bytes` should be private.

Reply:
> So we need a few things answered before we can solve this. Should undef bytes be treated as zero bytes here? Is bss cheaper than data? Only if we need all zeros anyway?

If some bits are 0 and some undefined, putting it into bss seems reasonable -- no reason to waste any effort on only initializing some of this with 0.

Reply:
if one byte is zero and 99999999 are undef that isn't true

Reply:
Can we ask the linker to zero some pages and leave the rest just uninitialized?

It is certainly always correct to replace undef bytes by 0, and the cost is tiny thanks to BSS and copy-on-write.

Reply:
You need to check for all zeros in the defined case, too